ryoanji/src/ryoanji/nbody/cartesian_qpole.hpp

```diff
@@ -72,19 +72,21 @@ struct Cqi
 
 /*! @brief Compute the monopole and quadruple moments from particle coordinates
  *
- * @tparam       T1             float or double
- * @tparam       T2             float or double
- * @tparam       T3             float or double
- * @param[in]    x              x coordinate array
- * @param[in]    y              y coordinate array
- * @param[in]    z              z coordinate array
- * @param[in]    m              masses array
- * @param[in]    numParticles   number of particles to read from coordinate arrays
- * @param[out]   gv             output quadrupole
+ * @tparam      T1     float or double
+ * @tparam      T2     float or double
+ * @tparam      T3     float or double
+ * @param[in]   x      x coordinate array
+ * @param[in]   y      y coordinate array
+ * @param[in]   z      z coordinate array
+ * @param[in]   m      masses array
+ * @param[in]   first  first particle to access in coordinate arrays
+ * @param[in]   last   last particle to access in coordinate arrays
+ * @param[in]   center center of mass of particles in input range
+ * @param[out]  gv     output quadrupole
  */
 template<class T1, class T2, class T3>
-HOST_DEVICE_FUN void particle2Multipole(const T1* x, const T1* y, const T1* z, const T2* m, LocalIndex first,
-                                        LocalIndex last, const Vec3<T1>& center, CartesianQuadrupole<T3>& gv)
+HOST_DEVICE_FUN void P2M(const T1* x, const T1* y, const T1* z, const T2* m, LocalIndex first, LocalIndex last,
```

Comment:
Change first/last to begin/end. This is consistent with M2M and doesn't imply that "end" is inclusive, as last does. This can be applied more generally throughout the codebase.

Reply:
done